src/react-apps/ux-editor/src/containers/Toolbar.tsx

```diff
@@ -34,109 +32,24 @@ export interface IToolbarState {
 }
 
 class ToolbarClass extends React.Component<IToolbarProps, IToolbarState> {
-  public toolbarComponents: IToolbarElement[] = [
-    {
-      label: HEADER,
-      actionMethod: () => {
-        FormActionDispatcher.addFormComponent({
-          component: HEADER,
-          title: HEADER,
-        }, (component, id) => {
-          this.handleNext(component, id);
-        },
-        );
-      },
-    },
-    {
-      label: TEXT_INPUT,
-      actionMethod: () => {
-        FormActionDispatcher.addFormComponent({
-          component: TEXT_INPUT,
-          title: TEXT_INPUT,
-        }, (component, id) => {
-          this.handleNext(component, id);
-        },
-        );
-      },
-    },
-    {
-      label: CHECKBOX,
-      actionMethod: () => {
-        FormActionDispatcher.addFormComponent({
-          component: CHECKBOX,
-          title: CHECKBOX,
-          options: [{ value: 'value', label: 'label' }],
-        }, (component, id) => {
-          this.handleNext(component, id);
-        },
-        );
-      },
-    },
-    {
-      label: TEXT_AREA,
-      actionMethod: () => {
-        FormActionDispatcher.addFormComponent({
-          component: TEXT_AREA,
-          title: TEXT_AREA,
-        }, (component, id) => {
-          this.handleNext(component, id);
-        },
-        );
-      },
-    },
-    {
-      label: RADIO_BUTTONS,
+  public toolbarComponents: IToolbarElement[] = components.map((c: any) => {
+    const customProperties = c.customProperties ? c.customProperties : {};
+    return {
+      label: c.name,
       actionMethod: () => {
         FormActionDispatcher.addFormComponent({
-          component: RADIO_BUTTONS,
-          title: RADIO_BUTTONS,
-          options: [{ value: 'value', label: 'label' }],
-        }, (component, id) => {
+          component: c.name,
+          itemType: LayoutItemType.Component,
+          title: c.name,
+          ...JSON.parse(JSON.stringify(customProperties)),
```

Comment:
Do you really need to JSON.stringify and then JSON.parse? Wouldn't that just do the same as just having:
`... customProperties` ?

Reply:
No. `...customProperties` adds the same object every time, and if you have multiple components of the same type (f.ex. dropdownlist), making changes to the custom properties will be reflected in all those components. 
Tried shallow copy via `Object.assign`, but that had the same results... Open to other solutions, but this was the only way I could get it working.